cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/TestCommand.java

```diff
@@ -336,19 +351,39 @@ public void execute() {
         // Check package files are modified after last build
         boolean isPackageModified = isProjectUpdated(project);
 
-        TaskExecutor taskExecutor = new TaskExecutor.TaskBuilder()
-                .addTask(new CleanTargetCacheDirTask(), isSingleFile) // clean the target cache dir(projects only)
-                .addTask(new ResolveMavenDependenciesTask(outStream)) // resolve maven dependencies in Ballerina.toml
+        CleanTargetCacheDirTask cleanTargetCacheDirTask = new CleanTargetCacheDirTask();
+        ResolveMavenDependenciesTask resolveMavenDependenciesTask = new ResolveMavenDependenciesTask(outStream);
+        CompileTask compileTask = new CompileTask(outStream, errStream, false, isPackageModified,
+                buildOptions.enableCache());
+        RunTestsTask runTestsTask = new RunTestsTask(outStream, errStream, rerunTests, groupList, disableGroupList,
+                testList, includes, coverageFormat, moduleMap, listGroups, excludes, cliArgs);
+
+        CreateTestExecutableTask createTestExecutableTask = null;
+
+        if (!project.buildOptions().cloud().isEmpty()) {
+            //if cloud flag is set, create the test executable
+            createTestExecutableTask = new CreateTestExecutableTask(outStream, this.output, runTestsTask);
+        }
+
+        RunNativeImageTestTask runNativeImageTestTask = new RunNativeImageTestTask(outStream, rerunTests, groupList,
+                disableGroupList, testList, includes, coverageFormat, moduleMap, listGroups);
+        DumpBuildTimeTask dumpBuildTimeTask = new DumpBuildTimeTask(outStream);
+
+        TaskExecutor.TaskBuilder taskBuilder = new TaskExecutor.TaskBuilder();
+
+        taskBuilder.addTask(cleanTargetCacheDirTask, isSingleFile) // clean the target cache dir(projects only)
+                .addTask(resolveMavenDependenciesTask) // resolve maven dependencies in Ballerina.toml
                 // compile the modules
-                .addTask(new CompileTask(outStream, errStream, false, isPackageModified, buildOptions.enableCache()))
```

Comment:
shall we only add the Executable task addition without changing the existing implementation?

Reply:
By the existing implementation do you mean how these tasks are added to the TaskBuilder? Like how the Constructors are called within the addTask method? 

The only issue is that, the order matters with this approach and currently CreateTestExecutableTask requires the reference to the RunTestsTask to invoke required methods. Since, RunTestsTask happens after the creation of the test executable, we cannot construct this object after the CreateTestExecutableTask using this approach. 

However, currently RunTestsTask's execute method returns right after invocation if the cloud flag is enabled since we have delegated the execution of the tests to the C2C side. Therefore we do not actually need to add the RunTestsTask to the TaskBuilder. Hence, we can call RunTestsTask's constructor within CreateTestExecutableTask's constructor. Doing so, would cause readability issues. Furthermore, we also need to consider the cases where cloud flag is not given (Normal bal test execution).

Reply:
After the discussion, I have made the relevant changes to reflect the previous implementation while including the current one. The change can be found in [c329e24](https://github.com/ballerina-platform/ballerina-lang/pull/42183/commits/c329e2498f275dc35210d6b116bdc585a851cf45) commit.